### PR TITLE
orca overview doc tf and pytorch section

### PR DESCRIPTION
**Pull request by yangw1234**

### CODE REVIEW

File: docs/readthedocs/Orca/Overview/distributed-training-inference.md [hunk 1]
```diff
@@ -6,37 +6,178 @@
 
 ### **1. Estimator**
 
-To perform distributed training and inference, the user can  first create an Orca `Estimator` from any standard (single-node) TensorFlow, Kera or PyTorch model, and then call `Estimator.fit` or `Estimator.predict`  methods (using the [data-parallel processing pipeline]() as input).
+To perform distributed training and inference, the user can first create an Orca `Estimator` from any standard (single-node) TensorFlow, Kera or PyTorch model, and then call `Estimator.fit` or `Estimator.predict`  methods (using the [data-parallel processing pipeline](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html) as input).
```
Comment (hkvision) [inline docs/readthedocs/Orca/Overview/distributed-training-inference.md:9]: use relative path for the link

Comment (yangw1234) [reply docs/readthedocs/Orca/Overview/distributed-training-inference.md]: fixed

**Comment by yangw1234** [comment]
https://tf1-doc.readthedocs.io/en/latest/doc/Orca/Overview/distributed-training-inference.html#

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 2]
```diff
@@ -6,37 +6,178 @@
 
 ### **1. Estimator**
 
-To perform distributed training and inference, the user can  first create an Orca `Estimator` from any standard (single-node) TensorFlow, Kera or PyTorch model, and then call `Estimator.fit` or `Estimator.predict`  methods (using the [data-parallel processing pipeline]() as input).
+To perform distributed training and inference, the user can first create an Orca `Estimator` from any standard (single-node) TensorFlow, Kera or PyTorch model, and then call `Estimator.fit` or `Estimator.predict`  methods (using the [data-parallel processing pipeline](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html) as input).
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:9]: use relative link, not absolute url

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 3]
```diff
@@ -6,37 +6,178 @@
...
 
-Under the hood, the Orca `Estimator` will replicate the model on each node in the cluster, feed the data partition (generated by the data-parallel processing pipeline) on each node to the local model replica, and synchronize model parameters using various *backend* technologies (such as *Horovod*, `tf.distribute.MirroredStrategy`, `torch.distributed`, or the parameter sync layer in [*BigDL*]()).
+Under the hood, the Orca `Estimator` will replicate the model on each node in the cluster, feed the data partition (generated by the data-parallel processing pipeline) on each node to the local model replica, and synchronize model parameters using various *backend* technologies (such as *Horovod*, `tf.distribute.MirroredStrategy`, `torch.distributed`, or the parameter sync layer in [*BigDL*](https://bigdl-project.github.io/master/)).
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:11]: use https://github.com/intel-analytics/BigDL

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 4]
```diff
@@ -6,37 +6,178 @@
...
 
 ### **2. TensorFlow/Keras Estimator**
 
 #### **2.1 TensorFlow 1.15 and Keras 2.3**
+
+There two ways to create a TensorFlow `Estimator` for users of TensorFlow 1.15. One is to create the Estimator from a low level graph and another from a keras model. Examples are as follows:
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:17]: There two ways to create an `Estimator` for TensorFlow 1.15, either from a low level computation graph or a Keras model.

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 5]
```diff
@@ -6,37 +6,178 @@
...
+
+Graph API:
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:19]: - **TensorFlow Computation Graph**

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 6]
```diff
@@ -6,37 +6,178 @@
...
+# define a metric
+acc = accuracy(logits, labels)
+
+# create an estimator using endpoints of the graph
+est = Estimator.from_graph(inputs=images,
+                           outputs=logits,
+                           labels=labels,
+                           loss=loss,
+                           optimizer=tf.train.AdamOptimizer(),
+                           metrics={"acc": acc})
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:38]: do we need metrics here?

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 7]
```diff
@@ -6,37 +6,178 @@
...
+```
+
+Keras API:
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:41]: - **Keras model**

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 8]
```diff
@@ -6,37 +6,178 @@
...
+```python
+model = create_keras_lenet_model()
+model.compile(optimizer=keras.optimizers.RMSprop(),
+              loss='sparse_categorical_crossentropy',
+              metrics=['accuracy'])
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:46]: Do we need metrics here?

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 9]
```diff
@@ -6,37 +6,178 @@
...
+        epochs=max_epoch)
+```
+The `data` argument in `fit` method can be a spark DataFrame, a XShards or a `tf.data.Dataset`. See the *data-parallel processing pipeline* [page](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html) for more details.
+
+Inference:
+```python
+predictions = est.predict(data=df,
+                          feature_cols=['image'])
+```
+The `data` argument in `predict` method can be a spark DataFrame or a XShards. See the *data-parallel processing pipeline* [page](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html) for more details.
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:67]: Combine the train and predict code blocks.

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 10]
```diff
@@ -6,37 +6,178 @@
...
+
+View the related [Python API doc]() <TODO: link to be added> for more details.
+
 #### **2.2 TensorFlow 2.x and Keras 2.4+**
 
+Users of TensorFlow 2.x can create a TensorFlow `Estimator` from a keras model. The keras model should be created, compiled and returned in a python function. For example:
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:73]: The user can create an `Estimator` for TensorFlow 2.x from a Keras model (using a _Model Creator Function_)

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 11]
```diff
@@ -6,37 +6,178 @@
...
+def model_creator(config):
+    model = create_keras_lenet_model()
+    model.compile(optimizer=keras.optimizers.RMSprop(),
+                  loss='sparse_categorical_crossentropy',
+                  metrics=['accuracy'])
+    return model
+est = Estimator.from_keras(model_creator=model_creator)
+```
+
+The `model_creator` argument should be a function that takes a `config` dictionary and returns a compiled keras model
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:85]: a compiled Keras model

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 12]
```diff
@@ -6,37 +6,178 @@
...
+                steps_per_epoch=total_size // batch_size)
+```
+The `data` argument in `fit` method can be a spark DataFrame, a XShards or a function that returns a `tf.data.Dataset`. See the *data-parallel processing pipeline* [page](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html) for more details.
+
+Inference:
+```python
+predictions = est.predict(data=df,
+                          feature_cols=['image'])
+```
+The `data` argument in `predict` method can be a spark DataFrame or a XShards. See the *data-parallel processing pipeline* [page](https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/Overview/data-parallel-processing.html) for more details.
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:106]: combine the fit and predict block

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md [hunk 13]
```diff
@@ -6,37 +6,178 @@
...
+est.predict(xshards)
+```
+The input to `predict` methods should be a *XShards* (each element needs to be a `{"feature": a numpy ndarray}`). See the *data-parallel processing pipeline* [page]() for more details. <TODO: we need to add Spark Dataframe support too>
+
+View the related [Python API doc]() <TODO: link to be added> for more details.
 
 **Using `torch.distributed` or *Horovod* backend**
 
-<TODO: add description for `torch.distributed` or *Horovod* support>
+Alternatively, users can create a `Estimator` using `torch.distributed` or *Horovod* backend by specifying the `backend` argument to be "torch_distributed" or "horovod". In this case, the `model` and `optimizer` should be wrapped in a function. For example:
```
Comment (jason-dai) [inline readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:146]: create a PyTorch `Estimator`

**Comment by yangw1234** [comment]
Hi Jason, do we need HorovodTrainer, or should keep it as an internal API as it is low level and very different than other estimators.

**Comment by jason-dai** [comment]
> Hi Jason, do we need HorovodTrainer, or should keep it as an internal API as it is low level and very different than other estimators.

Maybe leave it out for now.

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 6 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:38]: It does not feel complete if there are no accuracy report for a classifier.

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 8 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md:46]: It does not feel complete if there are no accuracy report for a classifier. 

Here is an example from tensorflow website overview page.

image

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 2 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 3 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 4 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: Fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 5 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: Fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 7 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: Fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 9 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: Fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 10 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 11 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 13 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: Fixed

File: readthedocs/source/doc/Orca/Overview/distributed-training-inference.md (hunk 12 above)
Comment (yangw1234) [reply readthedocs/source/doc/Orca/Overview/distributed-training-inference.md]: Fixed

**Comment by yangw1234** [comment]
I'll first merge it
